0005-ordered-quantification.md

```diff
@@ -0,0 +1,626 @@
+
+# Motivation
+[motivation]: #motivation
+
+Loops are notoriously difficult for Dafny users to verify, and coming up with the correct loop invariants to prove a loop correct 
+is particularly challenging. The recent addition of `for` loops was a good step to abstract away from raw `while` loops,
+but only supports consecutive integer indices. They still fall short of the expressiveness and economy of `for` or `foreach` loops in
+many mainstream programming languages, which generally iterate through the contents of a collection of some kind.
+It is a well-established best practice to avoid manual loop indexes wherever possible, as they force the assumption that datatypes
+support efficient random access, and are more likely to be used incorrectly by accident:
+
+```dafny
+// Before:
+method AllPairs(s: seq<nat>) returns (result: seq<(nat, nat)>) {
+  for i := 0 to |s| {
+    var left := s[i];
+    for j := 0 to |s| {
+      var right := s[i]; // Whoops!
+      result := result + (left, right);  
+    }
+  }
+}
+
+// After:
+method AllPairs(s: seq<nat>) returns (result: seq<(nat, nat)>) {
+  result := [];
+  foreach left <- s, right <- s {
+    result := result + (left, right);
+  }
+}
+
+// Or even better, as an expression:
+function method AllPairs(s: seq<nat>): seq<(nat, nat)> {
+  seq left <- s, right <- s :: (left, right)
+}
+```
+
+A more serious shortcoming is that there is currently no efficient way to iterate over the elements of a `set` at all.
+The best alternative is to use the assign-such-that operator and set subtraction, illustrated in the reference manual as follows
+([sec 10.5.2](https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#1052-sets)):
+
+```dafny
+// s is a set<int>
+var ss := s;
+while ss != {}
+{
+  var i: int :| i in ss;
+  ss := ss - {i};
+  print i, "\n";
+}
+```
+
+This is functionally correct, but in the current compiler and runtimes takes quadratic time.
+https://github.com/dafny-lang/dafny/issues/2062 contains a more in-depth exploration of this issue;
+the overall conclusion is that improving the situation through optimization is risky and expensive.
+
+By comparison, the equivalent `foreach` loop is:
+
+```dafny
+foreach i <- s {
+  print i, "\n";
+}
+```
+
+The runtime implementation of this loop can use the native iteration features of the target language.
+
+Going further, it is better to avoid writing any imperative loop at all where possible. Sticking to immutable data and
+expressions rather than statements allows logic to be used in specifications as well as implementation,
+and reduces the effort Dafny users have to spend on proving their code correct.
+The proposed sequence comprehension expression allows more
+logic that ultimately produces a sequence of values to be expressed as a value rather than a statement.
+Just producing a sequence of the values in the set above, sorted by the natural ordering of `int` values,
+is simple using a sequence comprehension expression: `seq i: int | i in s`.
```

Comment:
I'm not sure that it's worth having this, since users can simply write `SetToSeq(s)`.  And, once you remove this (keeping only `<-`), I think you don't need the notion of quantification ordering at all?  I expanded on this below.